File: examples/mcp_client.js
/**
 * MCP Client for Text-to-Image Generator
 * ======================================
 * 
 * This is a JavaScript client for interacting with the MCP server.
 * Use this to integrate image generation into your Node.js applications.
 */

class MCPClient {
    /**
     * Initialize the MCP client.
     * 
     * @param {string} baseUrl - Base URL of the MCP server
     */
    constructor(baseUrl = 'http://localhost:3123') {
        this.baseUrl = baseUrl.replace(/\/$/, '');
    }

    /**
     * Get available MCP tools.
     * 
     * @returns {Promise<Object>} Dictionary containing available tools
     */
    async listTools() {
        const response = await fetch(`${this.baseUrl}/mcp/messages`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ method: 'tools/list' })
        });
        
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        
        return response.json();
    }

    /**
     * Generate an image using MCP.
     * 
     * @param {string} prompt - Text description of the image to generate
     * @returns {Promise<Object>} Dictionary containing the generation result
     */
    async generateImage(prompt) {
        const response = await fetch(`${this.baseUrl}/mcp/messages`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                method: 'tools/call',
                params: {
                    name: 'generate-image',
                    arguments: { prompt }
                }
            })
        });
        
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        
        return response.json();
    }

    /**
     * Get image status using MCP.
     * 
     * @param {string} imageId - ID of the generated image
     * @returns {Promise<Object>} Dictionary containing the image status
     */
    async getImageStatus(imageId) {
        const response = await fetch(`${this.baseUrl}/mcp/messages`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                method: 'tools/call',
                params: {
                    name: 'get-image-status',
                    arguments: { imageId }
                }
            })
        });
        
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        
        return response.json();
    }

    /**
     * Extract image ID from generation result.
     * 
     * @param {Object} result - Result from generateImage method
     * @returns {string|null} Image ID if found, null otherwise
     */
    extractImageId(result) {
        const content = result.content?.[0]?.text || '';
        if (content.includes('Image ID:')) {
            return content.split('Image ID: ')[1].split('.')[0];
        }
        return null;
    }

    /**
     * Extract image URL from generation result.
     * 
     * @param {Object} result - Result from generateImage method
     * @returns {string|null} Image URL if found, null otherwise
     */
    extractImageUrl(result) {
        const content = result.content?.[0]?.text || '';
        if (content.includes('Image URL:')) {
            return content.split('Image URL: ')[1];
        }
        return null;
    }

    /**
     * Set up Server-Sent Events listener.
     * 
     * @param {Function} onMessage - Callback for message events
     * @param {Function} onError - Callback for error events
     * @returns {EventSource} The EventSource instance
     */
    setupSSE(onMessage, onError) {
        const eventSource = new EventSource(`${this.baseUrl}/mcp/sse`);
        
        eventSource.onmessage = function(event) {
            try {
                const data = JSON.parse(event.data);
                onMessage(data);
            } catch (error) {
                console.error('Error parsing SSE data:', error);
            }
        };
        
        eventSource.onerror = function(error) {
            onError(error);
        };
        
        return eventSource;
    }
}

/**
 * Example workflow demonstrating MCP usage.
 */
async function mcpWorkflowExample() {
    console.log('🚀 MCP Client Workflow Example');
    console.log('='.repeat(40));
    
    const mcp = new MCPClient();
    
    try {
        // Step 1: List available tools
        console.log('📋 Listing available tools...');
        const tools = await mcp.listTools();
        console.log(`✅ Found ${tools.tools?.length || 0} tools`);
        
        // Step 2: Generate an image
        const prompt = 'a magical forest with glowing mushrooms and fairy lights';
        console.log(`🎨 Generating image: '${prompt}'`);
        
        const result = await mcp.generateImage(prompt);
        console.log('✅ Image generation initiated');
        
        // Step 3: Extract image ID and URL
        const imageId = mcp.extractImageId(result);
        const imageUrl = mcp.extractImageUrl(result);
        
        if (imageId) {
            console.log(`🆔 Image ID: ${imageId}`);
            console.log(`🔗 Image URL: ${imageUrl}`);
            
            // Step 4: Check status
            console.log('⏳ Checking image status...');
            const statusResult = await mcp.getImageStatus(imageId);
            console.log('📊 Status:', statusResult);
            
        } else {
            console.log('❌ Could not extract image ID from result');
            console.log('Result:', result);
        }
    
    } catch (error) {
        console.error('❌ Error:', error.message);
    }
}

/**
 * Example of generating multiple images.
 */
async function batchGenerateExample() {
    console.log('\n🔄 Batch Generation Example');
    console.log('='.repeat(40));
    
    const mcp = new MCPClient();
    const prompts = [
        'a serene mountain lake at sunset',
        'a futuristic city with flying cars',
        'a cozy coffee shop interior'
    ];
    
    const results = [];
    
    for (let i = 0; i < prompts.length; i++) {
        const prompt = prompts[i];
        console.log(`\n🎨 Generating image ${i + 1}/${prompts.length}: '${prompt}'`);
        
        try {
            const result = await mcp.generateImage(prompt);
            const imageId = mcp.extractImageId(result);
            const imageUrl = mcp.extractImageUrl(result);
            
            results.push({
                prompt,
                imageId,
                imageUrl,
                result
            });
            
            console.log(`✅ Generated: ${imageId}`);
            console.log(`🔗 URL: ${imageUrl}`);
            
            // Small delay between requests
            await new Promise(resolve => setTimeout(resolve, 1000));
            
        } catch (error) {
            console.error(`❌ Failed to generate image ${i + 1}:`, error.message);
        }
    }
    
    console.log(`\n📊 Batch complete: ${results.length}/${prompts.length} images generated`);
    return results;
}

/**
 * Example of using Server-Sent Events.
 */
function sseExample() {
    console.log('\n📡 Server-Sent Events Example');
    console.log('='.repeat(40));
    
    const mcp = new MCPClient();
    
    const eventSource = mcp.setupSSE(
        (data) => {
            console.log('📨 SSE Message:', data);
        },
        (error) => {
            console.error('❌ SSE Error:', error);
        }
    );
    
    console.log('✅ SSE connection established');
    console.log('Press Ctrl+C to stop listening...');
    
    // Keep the connection alive for demonstration
    setTimeout(() => {
        eventSource.close();
        console.log('🔌 SSE connection closed');
    }, 30000); // Close after 30 seconds
}

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { MCPClient, mcpWorkflowExample, batchGenerateExample, sseExample };
}

// Run examples if this file is executed directly
if (typeof window === 'undefined') {
    // Node.js environment
    (async () => {
        await mcpWorkflowExample();
        await batchGenerateExample();
        sseExample();
    })();
} 

File: examples/mcp_client.py
#!/usr/bin/env python3
"""
MCP Client for Text-to-Image Generator
======================================

This is a Python client for interacting with the MCP server.
Use this to integrate image generation into your Python applications.
"""

import requests
import json
import time
from typing import Dict, Any, Optional, List

class MCPClient:
    """Client for interacting with the Text-to-Image MCP server."""
    
    def __init__(self, base_url: str = "http://localhost:3123"):
        """
        Initialize the MCP client.
        
        Args:
            base_url: Base URL of the MCP server
        """
        self.base_url = base_url.rstrip('/')
        self.session = requests.Session()
    
    def list_tools(self) -> Dict[str, Any]:
        """
        Get available MCP tools.
        
        Returns:
            Dictionary containing available tools
        """
        response = self.session.post(
            f"{self.base_url}/mcp/messages",
            json={"method": "tools/list"}
        )
        response.raise_for_status()
        return response.json()
    
    def generate_image(self, prompt: str) -> Dict[str, Any]:
        """
        Generate an image using MCP.
        
        Args:
            prompt: Text description of the image to generate
            
        Returns:
            Dictionary containing the generation result
        """
        response = self.session.post(
            f"{self.base_url}/mcp/messages",
            json={
                "method": "tools/call",
                "params": {
                    "name": "generate-image",
                    "arguments": {"prompt": prompt}
                }
            }
        )
        response.raise_for_status()
        return response.json()
    
    def get_image_status(self, image_id: str) -> Dict[str, Any]:
        """
        Get image status using MCP.
        
        Args:
            image_id: ID of the generated image
            
        Returns:
            Dictionary containing the image status
        """
        response = self.session.post(
            f"{self.base_url}/mcp/messages",
            json={
                "method": "tools/call",
                "params": {
                    "name": "get-image-status",
                    "arguments": {"imageId": image_id}
                }
            }
        )
        response.raise_for_status()
        return response.json()
    
    def extract_image_id(self, result: Dict[str, Any]) -> Optional[str]:
        """
        Extract image ID from generation result.
        
        Args:
            result: Result from generate_image method
            
        Returns:
            Image ID if found, None otherwise
        """
        content = result.get('content', [{}])[0].get('text', '')
        if 'Image ID:' in content:
            return content.split('Image ID: ')[1].split('.')[0]
        return None
    
    def extract_image_url(self, result: Dict[str, Any]) -> Optional[str]:
        """
        Extract image URL from generation result.
        
        Args:
            result: Result from generate_image method
            
        Returns:
            Image URL if found, None otherwise
        """
        content = result.get('content', [{}])[0].get('text', '')
        if 'Image URL:' in content:
            return content.split('Image URL: ')[1]
        return None

def mcp_workflow_example():
    """Example workflow demonstrating MCP usage."""
    print("🚀 MCP Client Workflow Example")
    print("=" * 40)
    
    # Initialize client
    mcp = MCPClient()
    
    try:
        # Step 1: List available tools
        print("📋 Listing available tools...")
        tools = mcp.list_tools()
        print(f"✅ Found {len(tools.get('tools', []))} tools")
        
        # Step 2: Generate an image
        prompt = "a magical forest with glowing mushrooms and fairy lights"
        print(f"🎨 Generating image: '{prompt}'")
        
        result = mcp.generate_image(prompt)
        print("✅ Image generation initiated")
        
        # Step 3: Extract image ID and URL
        image_id = mcp.extract_image_id(result)
        image_url = mcp.extract_image_url(result)
        
        if image_id:
            print(f"🆔 Image ID: {image_id}")
            print(f"🔗 Image URL: {image_url}")
            
            # Step 4: Check status
            print("⏳ Checking image status...")
            status_result = mcp.get_image_status(image_id)
            print("📊 Status:", status_result)
            
        else:
            print("❌ Could not extract image ID from result")
            print("Result:", result)
    
    except requests.exceptions.RequestException as e:
        print(f"❌ Network error: {e}")
    except Exception as e:
        print(f"❌ Error: {e}")

def batch_generate_example():
    """Example of generating multiple images."""
    print("\n🔄 Batch Generation Example")
    print("=" * 40)
    
    mcp = MCPClient()
    prompts = [
        "a serene mountain lake at sunset",
        "a futuristic city with flying cars",
        "a cozy coffee shop interior"
    ]
    
    results = []
    
    for i, prompt in enumerate(prompts, 1):
        print(f"\n🎨 Generating image {i}/{len(prompts)}: '{prompt}'")
        
        try:
            result = mcp.generate_image(prompt)
            image_id = mcp.extract_image_id(result)
            image_url = mcp.extract_image_url(result)
            
            results.append({
                'prompt': prompt,
                'image_id': image_id,
                'image_url': image_url,
                'result': result
            })
            
            print(f"✅ Generated: {image_id}")
            print(f"🔗 URL: {image_url}")
            
            # Small delay between requests
            time.sleep(1)
            
        except Exception as e:
            print(f"❌ Failed to generate image {i}: {e}")
    
    print(f"\n📊 Batch complete: {len(results)}/{len(prompts)} images generated")
    return results

if __name__ == "__main__":
    # Run the workflow example
    mcp_workflow_example()
    
    # Run the batch generation example
    batch_generate_example() 